### PR TITLE
Elasticsearch: Support for visualizing logs in Explore 

### LINKED ISSUE #16812: Support Elasticsearch for visualizing logs in Explore
Add Elasticsearch support for visualizing logs in Explore.

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
Support for visualizing logs in Explore using Elasticsearch.

Left to do for MVP:
- [x] Use field settings from datasource settings
- [x] Figure out the best way of sending query from query field to datasource.
- [x] Replace `context` property with something that says that the query is a logs query.
- [x] Logs query should be run when switching to logs mode even though it's empty.
- [x] Try use the `QueryField` without any highlighting or auto-complete to get styling and ctrl+Return/Return to run query. Need to look into why the input field is lagging when input text to it. If lag cannot be removed, lets go for a regular html input field instead with Return handler.
- [x] Add types for elasticsearch options to datasource in similar manner as Influx datasource

**Which issue(s) this PR fixes**:
Closes #16812 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by davkal** [comment]
Reduce scope for MVP:
- keep HTML input field for now, with Return handler
- clean up query handling, ambiguous query handling in the datasource, might need to probe on query type
- use field settings from datasource settings
- make sure empty (all) queries work, perhaps via `*`

6.3 beta is blocked by this.

**Comment by kaydelaney** [comment]
Re: Logs query should be run when switching to logs mode even though it's empty.

The simplest way to do this would probably be to add a dummy parameter to `ElasticsearchQuery`, that way when `hasNonEmptyQuery()` is called it fits the definition of being non-empty despite the query itself being an empty string. ("A target is non-empty when it has keys (with non-empty values) other than refId, key and context.")
Perhaps a bit hacky though

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/types.ts]: This should probably be optional.

File: public/app/plugins/datasource/elasticsearch/datasource.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/datasource.ts]: Awesome :+1:

**Review by marefr — COMMENTED** [review]
Typings of options looks great! Left some comments. One thing to consider is to copy the typings for the `query` function from `DataSourceApi<>` which now will give you typed queries/targets and proper Promise as return value.

**Comment by marefr** [comment]
> Re: Logs query should be run when switching to logs mode even though it's empty.
> 
> The simplest way to do this would probably be to add a dummy parameter to `ElasticsearchQuery`, that way when `hasNonEmptyQuery()` is called it fits the definition of being non-empty despite the query itself being an empty string. ("A target is non-empty when it has keys (with non-empty values) other than refId, key and context.")
> Perhaps a bit hacky though

@kaydelaney if we have a property on `ElasticsearchQuery` that says whether query is a logs query or not, that property can be added to hasNonEmptyQuery. What do you think?

**Comment by kaydelaney** [comment]
> > Re: Logs query should be run when switching to logs mode even though it's empty.
> > The simplest way to do this would probably be to add a dummy parameter to `ElasticsearchQuery`, that way when `hasNonEmptyQuery()` is called it fits the definition of being non-empty despite the query itself being an empty string. ("A target is non-empty when it has keys (with non-empty values) other than refId, key and context.")
> > Perhaps a bit hacky though
> 
> @kaydelaney if we have a property on `ElasticsearchQuery` that says whether query is a logs query or not, that property can be added to hasNonEmptyQuery. What do you think?

Sounds good. 👍

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/types.ts]: This should be used instead of the `ElasticsearchQuery` interface defined in `ElasticsearchQueryField.tsx`. Add missing fields from `ElasticsearchQueryField.tsx`.

File: public/app/plugins/datasource/elasticsearch/datasource.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/datasource.ts]: Think we can remove the `context` from the query now and only rely on `isLogsQuery` or does that screw up the check for empty query?

**Review by marefr — COMMENTED** [review]
Great progress :+1: Left some comments that needs to be fixed.

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (kaydelaney) [reply public/app/plugins/datasource/elasticsearch/types.ts]: So should the `ElasticsearchQuery` interface defined in `ElasticsearchQueryField.tsx` just be removed? (After adding the missing fields to the interface defined in `types.ts`)

Comment (marefr) [reply public/app/plugins/datasource/elasticsearch/types.ts]: Yes, there shouldn't be anyone using that any longer.

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/types.ts]: Is this still needed or can we remove it?

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/types.ts]: `metricAggs` was a typo from my side I think and should be removed.

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/types.ts]: Possible to remove `any` and make it typed using `ElasticsearchAggregation[]`?

File: public/app/plugins/datasource/elasticsearch/datasource.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/datasource.ts]: Remove comment :smile:

**Review by marefr — COMMENTED** [review]
Changes looks great. Added some comments with request for minor code changes and cleanup.

Two things that doesn't work as expected is using the `Clear All` and clear query field row. They are actually clearing current query correctly, but the query isn't run at all which is the first problem. 
Second, if you after clear click `Run Query` nothing happens.

Let me know if you'll need more information provided to reproduce these problems.

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (kaydelaney) [reply public/app/plugins/datasource/elasticsearch/types.ts]: No longer needed I think. Good catch

File: public/app/plugins/datasource/elasticsearch/datasource.ts
Comment (kaydelaney) [reply public/app/plugins/datasource/elasticsearch/datasource.ts]: Oops! 😅

File: public/app/plugins/datasource/elasticsearch/types.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/types.ts]: This can be removed since it not in use anymore

File: public/app/plugins/datasource/elasticsearch/specs/datasource.test.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/specs/datasource.test.ts]: Would be nice with a test for "When issuing logs query"

File: public/app/plugins/datasource/elasticsearch/query_builder.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/query_builder.ts]: Would be nice with a test of this verifying that both documents and date histogram is requested.

File: public/app/plugins/datasource/elasticsearch/elastic_response.ts
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/elastic_response.ts]: Would be nice with a test verifying that a response of both documents and date histogram aggregation can be properly parsed.

File: public/app/plugins/datasource/elasticsearch/components/ElasticsearchQueryField.tsx
Comment (marefr) [inline public/app/plugins/datasource/elasticsearch/components/ElasticsearchQueryField.tsx]: Remove commented code

**Review by marefr — COMMENTED** [review]
All features now works as expected :+1: Some minor cleanup and adding some tests would be good before we merge this, see comments..

**Review by marefr — COMMENTED** [review]
Great work! Approved. Since I'm the author I cannot approve this PR. @kaydelaney you have to approve it

**Review by kaydelaney — APPROVED** [review]

**Pull request merged**
